src/libmongoc/src/mongoc/mongoc-client.c

```diff
@@ -822,6 +825,97 @@ mongoc_client_connect (bool buffered,
    return base_stream;
 }
 
+#if defined(MONGOC_ENABLE_SSL_OPENSSL) && OPENSSL_VERSION_NUMBER >= 0x10100000L
+/*
+ *--------------------------------------------------------------------------
+ *
+ * mongoc_client_connect_with_openssl_context --
+ *
+ *       Connect to a MongoDB server using OpenSSL.
+ *
+ *       @ssl_ctx is the global OpenSSL context for the mongoc_client_t
+ *       associated with this function call.
+ *
+ * Returns:
+ *       A newly allocated mongoc_stream_t if successful; otherwise
+ *       NULL and @error is set.
+ *
+ * Side effects:
+ *       @error is set if return value is NULL.
+ *
+ *--------------------------------------------------------------------------
+ */
+mongoc_stream_t *
+mongoc_client_connect_with_openssl_context (bool buffered,
+                                            bool use_ssl,
+                                            void *ssl_opts_void,
+                                            SSL_CTX *ssl_ctx,
+                                            const mongoc_uri_t *uri,
+                                            const mongoc_host_list_t *host,
+                                            bson_error_t *error)
+{
```

Comment:
Suggest adding a `void *openssl_ctx_void` argument to `mongoc_client_connect` and moving the check of `MONGOC_ENABLE_SSL_OPENSSL` within the function body:

```c
mongoc_stream_t *
mongoc_client_connect (bool buffered,
                       bool use_ssl,
                       void *ssl_opts_void,
                       const mongoc_uri_t *uri,
                       const mongoc_host_list_t *host,
                       void *openssl_ctx_void, // New!
                       bson_error_t *error)
{
    // ... Same as before ...

#if defined(MONGOC_ENABLE_SSL_OPENSSL) && OPENSSL_VERSION_NUMBER >= 0x10100000L
         // Use shared OpenSSL context.
         base_stream =
            mongoc_stream_tls_new_with_hostname_and_openssl_context (base_stream, host->host, ssl_opts, true, (SSL_CTX*) openssl_ctx_void);
#else
         base_stream = mongoc_stream_tls_new_with_hostname (base_stream, host->host, ssl_opts, true);
#endif

    // ... Same as before ...
}
```

This follows a similar pattern to `ssl_opts_void` (which is similarly only defined if SSL is enabled). I expect this may reduce duplicate code.